File: app/src/lib/utils/keyCodes.ts
/**
 * Stable key-name vocabulary shared with the Rust `key_codes` module.
 *
 * The chord persistence layer stores rdev `Key` variant names ("MetaRight",
 * "AltGr", "KeyA", …) so the same array round-trips losslessly between
 * the picker UI, the SQLite settings row, and the global hotkey listener.
 *
 * This module owns the conversions between three vocabularies:
 *   - browser `KeyboardEvent` (`event.code` like "MetaRight" / "AltRight")
 *   - canonical chord key names (matches rdev variants)
 *   - human display labels ("⌘", "⌥", "A", …)
 */

/**
 * Map a `KeyboardEvent` to the canonical key name we persist. Returns
 * `null` for keys we don't support in chords (dead keys, IME composition,
 * etc.).
 *
 * Browser quirk: right-Option on macOS is reported as `"AltRight"`; rdev
 * calls it `"AltGr"`. Normalize to rdev's name so the Rust side recognizes
 * it without an aliasing layer.
 */
export function canonicalKeyFromEvent(event: KeyboardEvent): string | null {
  const code = event.code;
  if (!code) return null;
  switch (code) {
    case 'AltLeft':
      return 'Alt';
    case 'AltRight':
      return 'AltGr';
    case 'BracketLeft':
      return 'LeftBracket';
    case 'BracketRight':
      return 'RightBracket';
    case 'Semicolon':
      return 'SemiColon';
    case 'Backslash':
      return 'BackSlash';
    case 'Backquote':
      return 'BackQuote';
    case 'Period':
      return 'Dot';
    case 'Enter':
      return 'Return';
    case 'ArrowUp':
      return 'UpArrow';
    case 'ArrowDown':
      return 'DownArrow';
    case 'ArrowLeft':
      return 'LeftArrow';
    case 'ArrowRight':
      return 'RightArrow';
    default:
      // Browser names like "MetaRight", "MetaLeft", "ControlLeft",
      // "ShiftRight", "Space", "KeyA", "Digit1", "F5" all match the
      // rdev variant names directly.
      if (
        /^(Meta|Control|Shift)(Left|Right)$/.test(code) ||
        /^Key[A-Z]$/.test(code) ||
        /^Digit[0-9]$/.test(code) ||
        /^F([1-9]|1[0-2])$/.test(code) ||
        ['Space', 'Tab', 'Backspace', 'Delete', 'Escape', 'Insert',
          'Home', 'End', 'PageUp', 'PageDown', 'CapsLock', 'Function',
          'Minus', 'Equal', 'Quote', 'Comma', 'Slash'].includes(code)
      ) {
        return code;
      }
      return null;
  }
}

const PLATFORM_IS_MAC =
  typeof navigator !== 'undefined' && /mac/i.test(navigator.platform);

/**
 * Pretty label for a canonical key name. Picks platform-appropriate
 * modifier glyphs so macOS users see ⌘ and Windows/Linux users see Win.
 */
export function displayLabelForKey(name: string): string {
  switch (name) {
    case 'MetaLeft':
    case 'MetaRight':
      return PLATFORM_IS_MAC ? '⌘' : 'Win';
    case 'Alt':
      return PLATFORM_IS_MAC ? '⌥' : 'Alt';
    case 'AltGr':
      return PLATFORM_IS_MAC ? '⌥' : 'AltGr';
    case 'ControlLeft':
    case 'ControlRight':
      return PLATFORM_IS_MAC ? '⌃' : 'Ctrl';
    case 'ShiftLeft':
    case 'ShiftRight':
      return PLATFORM_IS_MAC ? '⇧' : 'Shift';
    case 'CapsLock':
      return '⇪';
    case 'Function':
      return 'fn';
    case 'Space':
      return 'Space';
    case 'Tab':
      return '⇥';
    case 'Return':
      return '↵';
    case 'Backspace':
      return '⌫';
    case 'Delete':
      return '⌦';
    case 'Escape':
      return 'Esc';
    case 'UpArrow':
      return '↑';
    case 'DownArrow':
      return '↓';
    case 'LeftArrow':
      return '←';
    case 'RightArrow':
      return '→';
  }
  if (/^Key([A-Z])$/.test(name)) return name.slice(3);
  if (/^Num([0-9])$/.test(name)) return name.slice(3);
  if (/^F([1-9]|1[0-2])$/.test(name)) return name;
  return name;
}

/**
 * Side-aware suffix to disambiguate left vs right modifier variants
 * — the tiny "R" badge that lets a user see the chord defaults to the
 * right-hand keys.
 */
export function modifierSideHint(name: string): 'L' | 'R' | null {
  if (name === 'MetaRight' || name === 'AltGr' || name === 'ControlRight' || name === 'ShiftRight') {
    return 'R';
  }
  if (name === 'MetaLeft' || name === 'Alt' || name === 'ControlLeft' || name === 'ShiftLeft') {
    return 'L';
  }
  return null;
}

/**
 * Sort a chord's keys so the kbd pills always render in a predictable
 * order: modifiers first (Ctrl, Opt, Shift, Cmd), main key last. Matches
 * how every macOS shortcut docs list the keys.
 */
const SORT_ORDER: Record<string, number> = {
  ControlLeft: 0, ControlRight: 0,
  Alt: 1, AltGr: 1,
  ShiftLeft: 2, ShiftRight: 2,
  MetaLeft: 3, MetaRight: 3,
  Function: 4,
  CapsLock: 5,
};

export function sortChordKeys(keys: string[]): string[] {
  return [...keys].sort((a, b) => {
    const sa = SORT_ORDER[a] ?? 99;
    const sb = SORT_ORDER[b] ?? 99;
    if (sa !== sb) return sa - sb;
    return a.localeCompare(b);
  });
}

File: tauri/src-tauri/src/key_codes.rs
//! Stable string ↔ `rdev::Key` mapping for chord persistence.
//!
//! The frontend captures keypresses through the browser keyboard API (which
//! exposes `event.code` like `"MetaRight"`, `"AltRight"`, `"Space"`, `"KeyA"`)
//! and stores chords in capture_settings as JSON arrays of canonical names.
//! On the way back the same names need to round-trip into `rdev::Key`
//! variants the chord engine actually matches against.
//!
//! Names follow the rdev variant identifiers exactly (`"MetaRight"`,
//! `"AltGr"`, `"KeyA"`, …) with one alias: the browser reports right-Option
//! as `"AltRight"` while rdev calls it `"AltGr"`. Both map to the same key.

use rdev::Key;

/// Resolve a canonical key name to its `rdev::Key`. Returns `None` for
/// names that don't have a corresponding variant — the command surface
/// rejects those so we never silently drop keys from a chord.
pub fn key_from_str(name: &str) -> Option<Key> {
    Some(match name {
        // Modifiers — left/right distinction matters for chord defaults.
        "Alt" | "AltLeft" => Key::Alt,
        "AltGr" | "AltRight" => Key::AltGr,
        "ControlLeft" => Key::ControlLeft,
        "ControlRight" => Key::ControlRight,
        "MetaLeft" => Key::MetaLeft,
        "MetaRight" => Key::MetaRight,
        "ShiftLeft" => Key::ShiftLeft,
        "ShiftRight" => Key::ShiftRight,
        "CapsLock" => Key::CapsLock,
        "Function" => Key::Function,

        // Whitespace / navigation
        "Space" => Key::Space,
        "Tab" => Key::Tab,
        "Return" | "Enter" => Key::Return,
        "Backspace" => Key::Backspace,
        "Delete" => Key::Delete,
        "Escape" => Key::Escape,
        "Insert" => Key::Insert,
        "Home" => Key::Home,
        "End" => Key::End,
        "PageUp" => Key::PageUp,
        "PageDown" => Key::PageDown,
        "ArrowUp" | "UpArrow" => Key::UpArrow,
        "ArrowDown" | "DownArrow" => Key::DownArrow,
        "ArrowLeft" | "LeftArrow" => Key::LeftArrow,
        "ArrowRight" | "RightArrow" => Key::RightArrow,

        // Function row
        "F1" => Key::F1, "F2" => Key::F2, "F3" => Key::F3, "F4" => Key::F4,
        "F5" => Key::F5, "F6" => Key::F6, "F7" => Key::F7, "F8" => Key::F8,
        "F9" => Key::F9, "F10" => Key::F10, "F11" => Key::F11, "F12" => Key::F12,

        // Digits
        "Digit0" | "Num0" => Key::Num0,
        "Digit1" | "Num1" => Key::Num1,
        "Digit2" | "Num2" => Key::Num2,
        "Digit3" | "Num3" => Key::Num3,
        "Digit4" | "Num4" => Key::Num4,
        "Digit5" | "Num5" => Key::Num5,
        "Digit6" | "Num6" => Key::Num6,
        "Digit7" | "Num7" => Key::Num7,
        "Digit8" | "Num8" => Key::Num8,
        "Digit9" | "Num9" => Key::Num9,

        // Letters — browser uses "KeyA" style which already matches rdev.
        "KeyA" => Key::KeyA, "KeyB" => Key::KeyB, "KeyC" => Key::KeyC,
        "KeyD" => Key::KeyD, "KeyE" => Key::KeyE, "KeyF" => Key::KeyF,
        "KeyG" => Key::KeyG, "KeyH" => Key::KeyH, "KeyI" => Key::KeyI,
        "KeyJ" => Key::KeyJ, "KeyK" => Key::KeyK, "KeyL" => Key::KeyL,
        "KeyM" => Key::KeyM, "KeyN" => Key::KeyN, "KeyO" => Key::KeyO,
        "KeyP" => Key::KeyP, "KeyQ" => Key::KeyQ, "KeyR" => Key::KeyR,
        "KeyS" => Key::KeyS, "KeyT" => Key::KeyT, "KeyU" => Key::KeyU,
        "KeyV" => Key::KeyV, "KeyW" => Key::KeyW, "KeyX" => Key::KeyX,
        "KeyY" => Key::KeyY, "KeyZ" => Key::KeyZ,

        // Punctuation / symbols
        "Backquote" | "BackQuote" => Key::BackQuote,
        "Minus" => Key::Minus,
        "Equal" => Key::Equal,
        "BracketLeft" | "LeftBracket" => Key::LeftBracket,
        "BracketRight" | "RightBracket" => Key::RightBracket,
        "Semicolon" | "SemiColon" => Key::SemiColon,
        "Quote" => Key::Quote,
        "Backslash" | "BackSlash" => Key::BackSlash,
        "Comma" => Key::Comma,
        "Period" | "Dot" => Key::Dot,
        "Slash" => Key::Slash,

        _ => return None,
    })
}
